Make DispatcherTimer implement IDisposable.

grokys opened the pull request:
Fixes #1399 

cc: @lindexi

grokys committed:
Make DispatcherTimer implement IDisposable.

Fixes #1399

aelij commented:
Hmm, actually, after reading the entire class, I see it has a finalized, and I think the finalizer should also call Dispose. The proper way to implement IDisposable with a finalized is to have a private/protected Dispose(bool).

aelij commented on src/Avalonia.Base/Threading/DispatcherTimer.cs:
```diff
@@ -117,6 +117,11 @@ public object Tag
+        /// <summary>
+        /// Stops the timer and disposes of all unmanaged resources.
+        /// </summary>
+        public void Dispose() => _timer?.Dispose();
```
Add a Dispose(bool) method and have Dispose() and the finalizer call it (passing in true and false, respectively)

aelij replied:
Also add GC.SupressFinalize(this) in Dispose()

aelij commented:
Another thing, why does this class even have a finalizer? It only holds managed resources. A finalizer is usually only needed in a class holding unmanaged resources.

grokys commented:
> Another thing, why does this class even have a finalizer? It only holds managed resources.

The `IDisposable _timer` member represents an unmanaged resource: the underlying OS timer. Disposing this calls for example `KillTimer(IntPtr.Zero, handle);` on Win32.

You're right about using the disposable pattern though, will do that.

grokys committed:
Use dispose pattern on DispatcherTimer.

Though don't call `GC.SuppressFinalize` because the timer could be resurrected.

grokys commented:
I've implemented the dispose pattern on `DispatcherTimer` except the `GC.SuppressFinalize(this)` part. This is because disposing a `DispatcherTimer` just stops the timer: it could be restarted. If we want to prevent this we'd need to add an extra `bool disposed` member which seems a bit wasteful. What do you think?

grokys committed:
Merge branch 'master' into fixes/1399-dispatchertimer-disposable

aelij commented:
> The IDisposable _timer member represents an unmanaged resource

It doesn't directly. `IDisposable` is **necessarily** a managed object. Only real unmanaged resources (such as `IntPtr`s) should have finalizers. Having unnecessary objects in the finalization queue is a GC perf issue.

In MonoMac, for example, `StartTimer` returns an NSTimer which inherits from [NSObject which has a finalizer](https://developer.xamarin.com/api/member/MonoTouch.Foundation.NSObject.Finalize()/).

In Win32, you should be using a `SafeHandle` (which is a `CriticalFinalizerObject`) instead of raw `IntPtr`s - otherwise if someone used `IPlatformThreadingInterface` directly, there could be a resource leak. `Disposable.Create` with `KillTimer` is not good enough here:

https://github.com/AvaloniaUI/Avalonia/blob/5a90c5f0a8914ba3c3d603b2a39d913076b2b4ec/src/Windows/Avalonia.Win32/Win32Platform.cs#L132-L136

So `DispatcherTimer` should only have a `Dispose` without a finalizer that disposes the platform timer.

kekekeks commented:
>The IDisposable _timer member represents an unmanaged resource

The underlying timer should be a SafeHandle or have a finalizer. The rule of the thumb here is that if the class doesn't have an IntPtr, it shouldn't have a finalizer.

grokys committed:
Use a SafeHandle for timers.

grokys commented:
Thanks for the guidance on this @aelij @kekekeks, I'm not so familiar with this stuff. Now using a `SafeHandle` - does that look right to you?

grokys committed:
Merge branch 'master' into fixes/1399-dispatchertimer-disposable

aelij commented on src/Windows/Avalonia.Win32/TimerHandle.cs:
```diff
@@ -0,0 +1,23 @@
+
+namespace Avalonia.Win32
+{
+    internal class TimerHandle : CriticalHandleZeroOrMinusOneIsInvalid
```
Why `CriticalHandle` rather than `SafeHandle`?

grokys replied:
Ah, that was a mistake. I meant to use `SafeHandleZeroOrMinusOneIsInvalid`, not sure how I ended up with the wrong one. Will fix.

grokys committed:
Use SafeHandle not CriticalHandle.

grokys committed:
Merge branch 'master' into fixes/1399-dispatchertimer-disposable

grokys commented:
@aelij now using `SafeHandle` instead of `CriticalHandle`.

aelij commented on src/Avalonia.Base/Threading/DispatcherTimer.cs:
```diff
@@ -200,7 +192,10 @@ public void Stop()
         }
 
-        
+        protected virtual void Dispose(bool disposing)
```
There's no need for the `Dispose(bool)` pattern when there's no finalizer. Just make `Dispose()` virtual if needed.

aelij commented on src/Windows/Avalonia.Win32/Win32Platform.cs:
```diff
@@ -126,14 +125,7 @@ public IDisposable StartTimer(DispatcherPriority priority, TimeSpan interval, Ac
                 timerDelegate);
 
-            // Prevent timerDelegate being garbage collected.
-            _delegates.Add(timerDelegate);
```
The `SafeHandle` is susceptible to GC as well. Is it enough that it's being held by the `DispatcherTimer`? Why was this delegate array needed in the first place?

Let's say I do `new DispatcherTimer()` and not hold onto its reference. I might expect it to continue ticking... This is how it would've behaved in WPF since timers were rooted by the Dispatcher, I think. `System.Timers.Timer` and `System.Threading.Timer` also behave this way.

lindexi commented:
@aelij Any news?
